File: crates/experimentation-hash/src/murmur3.rs
//! MurmurHash3 x86 32-bit implementation.
//!
//! Reference: https://github.com/aappleby/smhasher/blob/master/src/MurmurHash3.cpp
//!
//! This implementation uses LITTLE-ENDIAN byte order.
//! WASM also runs little-endian, so results are consistent across targets.

const C1: u32 = 0xcc9e_2d51;
const C2: u32 = 0x1b87_3593;

/// MurmurHash3 x86 32-bit hash function.
///
/// This is the reference implementation. All other targets (WASM, FFI, UniFFI, PyO3)
/// must produce identical output for the same inputs.
pub fn murmurhash3_x86_32(data: &[u8], seed: u32) -> u32 {
    let len = data.len();
    let n_blocks = len / 4;
    let mut h1 = seed;

    // Body: process 4-byte blocks
    for i in 0..n_blocks {
        let offset = i * 4;
        let k1 = u32::from_le_bytes([
            data[offset],
            data[offset + 1],
            data[offset + 2],
            data[offset + 3],
        ]);

        let k1 = k1.wrapping_mul(C1);
        let k1 = k1.rotate_left(15);
        let k1 = k1.wrapping_mul(C2);

        h1 ^= k1;
        h1 = h1.rotate_left(13);
        h1 = h1.wrapping_mul(5).wrapping_add(0xe654_6b64);
    }

    // Tail: process remaining bytes
    let tail = &data[n_blocks * 4..];
    let mut k1: u32 = 0;

    match tail.len() {
        3 => {
            k1 ^= (tail[2] as u32) << 16;
            k1 ^= (tail[1] as u32) << 8;
            k1 ^= tail[0] as u32;
            k1 = k1.wrapping_mul(C1);
            k1 = k1.rotate_left(15);
            k1 = k1.wrapping_mul(C2);
            h1 ^= k1;
        }
        2 => {
            k1 ^= (tail[1] as u32) << 8;
            k1 ^= tail[0] as u32;
            k1 = k1.wrapping_mul(C1);
            k1 = k1.rotate_left(15);
            k1 = k1.wrapping_mul(C2);
            h1 ^= k1;
        }
        1 => {
            k1 ^= tail[0] as u32;
            k1 = k1.wrapping_mul(C1);
            k1 = k1.rotate_left(15);
            k1 = k1.wrapping_mul(C2);
            h1 ^= k1;
        }
        _ => {}
    }

    // Finalization mix
    h1 ^= len as u32;
    h1 = fmix32(h1);

    h1
}

#[inline]
fn fmix32(mut h: u32) -> u32 {
    h ^= h >> 16;
    h = h.wrapping_mul(0x85eb_ca6b);
    h ^= h >> 13;
    h = h.wrapping_mul(0xc2b2_ae35);
    h ^= h >> 16;
    h
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_known_values() {
        // Empty string with seed 0
        assert_eq!(murmurhash3_x86_32(b"", 0), 0);

        // Known test vectors from the reference implementation
        assert_eq!(murmurhash3_x86_32(b"hello", 0), 0x248b_fa47);
        assert_eq!(murmurhash3_x86_32(b"hello", 1), 0xbb4a_bccb);
    }

    #[test]
    fn test_deterministic() {
        let h1 = murmurhash3_x86_32(b"test_input", 42);
        let h2 = murmurhash3_x86_32(b"test_input", 42);
        assert_eq!(h1, h2);
    }
}

File: services/flags/internal/hash/murmur3.go
package hash

import (
	"encoding/binary"
	"math/bits"
)

const (
	c1 uint32 = 0xcc9e2d51
	c2 uint32 = 0x1b873593
)

// MurmurHash3X86_32 computes the MurmurHash3 x86 32-bit hash.
// This is a direct port of the Rust implementation at
// crates/experimentation-hash/src/murmur3.rs.
func MurmurHash3X86_32(data []byte, seed uint32) uint32 {
	length := len(data)
	nBlocks := length / 4
	h1 := seed

	// Body: process 4-byte blocks.
	for i := 0; i < nBlocks; i++ {
		offset := i * 4
		k1 := binary.LittleEndian.Uint32(data[offset : offset+4])

		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2

		h1 ^= k1
		h1 = bits.RotateLeft32(h1, 13)
		h1 = h1*5 + 0xe6546b64
	}

	// Tail: process remaining bytes.
	tail := data[nBlocks*4:]
	var k1 uint32

	switch len(tail) {
	case 3:
		k1 ^= uint32(tail[2]) << 16
		k1 ^= uint32(tail[1]) << 8
		k1 ^= uint32(tail[0])
		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2
		h1 ^= k1
	case 2:
		k1 ^= uint32(tail[1]) << 8
		k1 ^= uint32(tail[0])
		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2
		h1 ^= k1
	case 1:
		k1 ^= uint32(tail[0])
		k1 *= c1
		k1 = bits.RotateLeft32(k1, 15)
		k1 *= c2
		h1 ^= k1
	}

	// Finalization mix.
	h1 ^= uint32(length)
	h1 = fmix32(h1)

	return h1
}

func fmix32(h uint32) uint32 {
	h ^= h >> 16
	h *= 0x85ebca6b
	h ^= h >> 13
	h *= 0xc2b2ae35
	h ^= h >> 16
	return h
}
